src/mscorlib/src/System/Runtime/CompilerServices/ConditionalWeakTable.cs

```diff
@@ -150,6 +150,41 @@ public void Add(TKey key, TValue value)
         }
 
         //--------------------------------------------------------------------------------------------
+        // key: key to add or update. May not be null.
+        // value: value to associate with key.
+        //
+        // If the key is already entered into the dictionary, this method will update the value associated with key.
+        //--------------------------------------------------------------------------------------------
+        [System.Security.SecurityCritical]
+        public void AddOrUpdate(TKey key, TValue value)
+        {
+            if (key == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.key);
+            }
+
+            lock (_lock)
+            {
+                VerifyIntegrity();
+                _invalid = true;
+                int entryIndex = FindEntry(key);
+
+                // if we found a key we should just update, if no we should create a new entry.
+                if (entryIndex != -1)
+                {
+                    _entries[entryIndex].depHnd.Free();
+                    _entries[entryIndex].depHnd = new DependentHandle(key, value);
+                }
+                else
+                {
+                    CreateEntry(key, value);
```

Comment:
Do you want me to wait until @stephentoub ports the CoreRT improvements back to CoreCLR on issue #3206 so that I can do this API addition with no locks?